radiomesh/gridding.py

```diff
@@ -270,20 +278,17 @@ def apply_w_impl(
           y_kernel = eval_es_kernel(KERNEL, v_grid, v_pixel_start)
           z_kernel = eval_es_kernel(KERNEL, w_grid, w_pixel_start)
 
-          for zfi, zkw in zip(
-            numba.literal_unroll(z_indices), numba.literal_unroll(z_kernel)
-          ):
-            zi = int(zfi)
-            for xfi, xkw in zip(
-              numba.literal_unroll(x_indices), numba.literal_unroll(x_kernel)
-            ):
-              xi = int(xfi)
-              for yfi, ykw in zip(
-                numba.literal_unroll(y_indices), numba.literal_unroll(y_kernel)
-              ):
-                yi = int(yfi)
-                weighted_stokes = apply_weights(vis, xkw * ykw * zkw)
-                accumulate_data(weighted_stokes, vis_grid, (zi, xi, yi), 0)
+          for d in range(ndir):
+            didx = idx + (d,)
+            dir_vis = maybe_apply_jones(JONES_VIS_DATUM, jones_params, vis, didx)
+            dir_weight = maybe_apply_jones(JONES_WGT_DATUM, jones_params, wgt, didx)
+            dir_vis = apply_weights(dir_vis, dir_weight)
+
+            for zi, zkw in zip(literal_unroll(z_indices), literal_unroll(z_kernel)):
+              for xi, xkw in zip(literal_unroll(x_indices), literal_unroll(x_kernel)):
+                for yi, ykw in zip(literal_unroll(y_indices), literal_unroll(y_kernel)):
+                  weighted_stokes = apply_weights(dir_vis, xkw * ykw * zkw)
+                  accumulate_data(weighted_stokes, vis_grid, (zi, xi, yi), 0)
```

Comment:
@landmanbester @JSKenyon @bennahugo  Looking at this code block I again still think it should be possible to accumulate the per-direction visibilities (`dir_vis`) into `dir_vis_sum` and then just grid `dir_vis_sum` as the kernel weights do not change per direction.

I think this is equivalent to the above code block since `accumulate_data` is adding `dir_vis` to the same locations with the same kernel weights. Thus I think it's performing the same operation with more instructions, compared to the approach described in the previous paragraph.

This should save some memory accesses. I should probably test this hypothesis tomorrow.

Reply:
I could see this not working if the kernel for w changes per direction (i.e. in the faceting case)

Reply:
I will  defer to @landmanbester here - I think that you only apply one DD correction per grid though (assuming we are talking about faceting).

Reply:
> I think this is equivalent to the above code block since accumulate_data is adding dir_vis to the same locations with the same kernel weights. Thus I think it's performing the same operation with more instructions, compared to the approach described in the previous paragraph.

- Some hacking around in #45 seems to confirm the above. I assume gridding  the accumulated visibilities over direction probably  doesn't work in more complicated cases (faceting).

Reply:
> The kernel weights will be the same for each direction but vis_grid needs a direction axis. You can't grid all directions together when applying different jones terms per direction

Ahhhh, this is what I was missing conceptually. Do you have any strong opinions on where the `NDIR` axis should go in the current shape:  `(NSTOKES, NW, NX, NY)`?

Reply:
> This shouldn't have any effect on the images that are produced but it will alter the grid. What was failing before you introduced the conjugate?

IIRC, I moved the conjugation in `wgrid_data` which caused it to not agree with `explicit_gridder`. Changing `explicit_gridder`  then snowballed into changing `grid_data`

Reply:
> IIRC, I moved the conjugation in wgrid_data which caused it to not agree with explicit_gridder. Changing explicit_gridder then snowballed into changing grid_data

Strange but as long as we have agreement with the DFT all should be good.

> Ahhhh, this is what I was missing conceptually (and alluded to by @JSKenyon) . Do you have any strong opinions on where the NDIR axis should go in the current vis_grid shape: (NSTOKES, NW, NX, NY)?

Probably between the Stokes and w axes unless you have a good reason to put it elsewhere. Do you want me to modify the explicit gridder to incorporate a direction axis?

Reply:
> Probably between the Stokes and w axes unless you have a good reason to put it elsewhere.

done

> Do you want me to modify the explicit gridder to incorporate a direction axis?

yes, perhaps that can be left to another PR? I'm keen to put a fork in this one.